core/primitives/src/errors.rs

```diff
@@ -233,14 +233,24 @@ impl Display for BalanceMismatchError {
     }
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Debug, Clone, PartialEq, Eq)]
+pub struct IntegerOverflowError;
+
 /// Error returned from `Runtime::apply`
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub enum RuntimeError {
+    UnexpectedIntegerOverflow,
     InvalidTxError(InvalidTxError),
     StorageError(StorageError),
     BalanceMismatch(BalanceMismatchError),
 }
 
+impl From<IntegerOverflowError> for InvalidTxError {
```

Comment:
Actually it was the whole idea to make a cast in order to cast, depending on the returning type. Why you think it's important to be explicit here?

Reply:
As an alternative I can make a thing like a GasCounter in vm-logic. Two specific counters: one counts Tx gas and costs (incapsulating the corresponding logic and checks) and the second counts the cost of the (actually I don't know how to call the second one - tell me if you have ideas).

Reply:
I'm sorry. I meant `map_err` instead. When you have a conversion to `InvalidTxError`, there is a path towards `RuntimeError` through `InvalidTxError`. So it's unclear if `?` would map through the chain of `into` and `from` automatically. This would cause uncertainty whether it maps into `RuntimeError::InvalidTxError` or `RuntimeError::UnexpectedIntegerOverflow`.

Reply:
The current idea is the following: if the function where `IntegerOverflowError` returns `InvalidTxError` it will be converted to `InvalidTxError::CostOverflow`. If `IntegerOverflowError` happens in function which returns `RuntimeError` it will be converted into `RuntimeError::UnexpectedIntegerOverflow`. This a default behavior. In exceptional cases we make a cast.